### PR TITLE
[TypeDeclaration] Handle crash on intersection in Union type on AddMethodCallBasedStrictParamTypeRector

### LINKED ISSUE #7311: `Object of class PhpParser\\Node\\IntersectionType could not be converted to string` with `TYPE_DECLARATION_STRICT` set
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.13.9             |

Code and error is in this demo link: https://getrector.org/demo/88bae302-e888-4208-8485-f2f3acaf19dd

<details>
<summary>Here is the output if I run with `--debug`</summary>

```
[ERROR] Could not process "app/Payone/Requests/BaseRequest.php" file, due to:                                          
         "System error: "Object of class PhpParser\Node\IntersectionType could not be converted to string"              
                                                                                                                        
         Stack trace:                                                                                                   
         #0 vendor/rector/rector/packages/PHPStanStaticTypeMapper/TypeMapper/UnionTypeMapper.php(284): array_unique(Array)
         #1 vendor/rector/rector/packages/PHPStanStaticTypeMapper/TypeMapper/UnionTypeMapper.php(211):                  
         Rector\PHPStanStaticTypeMapper\TypeMapper\UnionTypeMapper->matchPhpParserUnionType(Object(PHPStan\Type\UnionType),
         'param')                                                                                                       
         #2 vendor/rector/rector/packages/PHPStanStaticTypeMapper/TypeMapper/UnionTypeMapper.php(139):                  
         Rector\PHPStanStaticTypeMapper\TypeMapper\UnionTypeMapper->matchTypeForUnionedObjectTypes(Object(PHPStan\Type\UnionType
         ), 'param')                                                                                                    
         #3 vendor/rector/rector/packages/PHPStanStaticTypeMapper/PHPStanStaticTypeMapper.php(63):                      
         Rector\PHPStanStaticTypeMapper\TypeMapper\UnionTypeMapper->mapToPhpParserNode(Object(PHPStan\Type\UnionType), 'param')
         #4 vendor/rector/rector/packages/StaticTypeMapper/StaticTypeMapper.php(85):                                    
         Rector\PHPStanStaticTypeMapper\PHPStanStaticTypeMapper->mapToPhpParserNode(Object(PHPStan\Type\UnionType), 'param')
         #5 vendor/rector/rector/rules/TypeDeclaration/NodeAnalyzer/ClassMethodParamTypeCompleter.php(61):              
         Rector\StaticTypeMapper\StaticTypeMapper->mapPHPStanTypeToPhpParserNode(Object(PHPStan\Type\UnionType), 'param')
         #6 vendor/rector/rector/rules/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector.php(96):
         Rector\TypeDeclaration\NodeAnalyzer\ClassMethodParamTypeCompleter->complete(Object(PhpParser\Node\Stmt\ClassMethod),
         Array, 3)                                                                                                      
         #7 vendor/rector/rector/src/Rector/AbstractRector.php(204):                                                    
         Rector\TypeDeclaration\Rector\ClassMethod\AddMethodCallBasedStrictParamTypeRector->refactor(Object(PhpParser\Node\Stmt</>
         ClassMethod))                                                                                                  
         #8 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(176):                          
         Rector\Core\Rector\AbstractRector->enterNode(Object(PhpParser\Node\Stmt\ClassMethod))                          
         #9 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105):                          
         PhpParser\NodeTraverser->traverseArray(Array)                                                                  
         #10 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196):                         
         PhpParser\NodeTraverser->traverseNode(Object(PhpParser\Node\Stmt\Class_))                                      
         #11 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(105):                         
         PhpParser\NodeTraverser->traverseArray(Array)                                                                  
         #12 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(196):                         
         PhpParser\NodeTraverser->traverseNode(Object(PhpParser\Node\Stmt\Namespace_))                                  
         #13 vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(85):                          
         PhpParser\NodeTraverser->traverseArray(Array)                                                                  
         #14 vendor/rector/rector/src/PhpParser/NodeTraverser/RectorNodeTraverser.php(42):                              
         PhpParser\NodeTraverser->traverse(Array)                                                                       
         #15 vendor/rector/rector/src/Application/FileProcessor.php(61):                                                
         Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser->traverse(Array)                                       
         #16 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(126):                              
         Rector\Core\Application\FileProcessor->refactor(Object(Rector\Core\ValueObject\Application\File),              
         Object(Rector\Core\ValueObject\Configuration))                                                                 
         #17 vendor/rector/rector/src/Application/FileProcessor/PhpFileProcessor.php(93):                               
         Rector\Core\Application\FileProcessor\PhpFileProcessor->refactorNodesWithRectors(Object(Rector\Core\ValueObject\Applica
         tion\File), Object(Rector\Core\ValueObject\Configuration))                                                     
         #18 vendor/rector/rector/packages/Parallel/WorkerRunner.php(99):                                               
         Rector\Core\Application\FileProcessor\PhpFileProcessor->process(Object(Rector\Core\ValueObject\Application\File),
         Object(Rector\Core\ValueObject\Configuration))                                                                 
         #19 vendor/rector/rector/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(97):                   
         Rector\Parallel\WorkerRunner->Rector\Parallel\{closure}(Array)                                                 
         #20 vendor/rector/rector/vendor/clue/ndjson-react/src/Decoder.php(110):                                        
         RectorPrefix202207\Evenement\EventEmitter->emit('data', Array)                                                 
         #21 vendor/rector/rector/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(97):                   
         RectorPrefix202207\Clue\React\NDJson\Decoder->handleData(Array)                                                
         #22 vendor/rector/rector/vendor/react/stream/src/Util.php(62): RectorPrefix202207\Evenement\EventEmitter->emit('data',
         Array)                                                                                                         
         #23 vendor/rector/rector/vendor/evenement/evenement/src/Evenement/EventEmitterTrait.php(97):                   
         RectorPrefix202207\React\Stream\Util::RectorPrefix202207\React\Stream\{closure}('{"action":"main...')          
         #24 vendor/rector/rector/vendor/react/stream/src/DuplexResourceStream.php(154):                                
         RectorPrefix202207\Evenement\EventEmitter->emit('data', Array)                                                 
         #25 vendor/rector/rector/vendor/react/event-loop/src/StreamSelectLoop.php(201):                                
         RectorPrefix202207\React\Stream\DuplexResourceStream->handleData(Resource id #3548)                            
         #26 vendor/rector/rector/vendor/react/event-loop/src/StreamSelectLoop.php(173):                                
         RectorPrefix202207\React\EventLoop\StreamSelectLoop->waitForStreamActivity(NULL)                               
         #27 vendor/rector/rector/src/Console/Command/WorkerCommand.php(63):                                            
         RectorPrefix202207\React\EventLoop\StreamSelectLoop->run()                                                     
         #28 vendor/rector/rector/vendor/symfony/console/Command/Command.php(307):                                      
         Rector\Core\Console\Command\WorkerCommand->execute(Object(RectorPrefix202207\Symfony\Component\Console\Input\ArgvInput)
         , Object(RectorPrefix202207\Symfony\Component\Console\Output\ConsoleOutput))                                   
         #29 vendor/rector/rector/vendor/symfony/console/Application.php(891):                                          
         RectorPrefix202207\Symfony\Component\Console\Command\Command->run(Object(RectorPrefix202207\Symfony\Component\Console\I
         nput\ArgvInput), Object(RectorPrefix202207\Symfony\Component\Console\Output\ConsoleOutput))                    
         #30 vendor/rector/rector/vendor/symfony/console/Application.php(310):                                          
         RectorPrefix202207\Symfony\Component\Console\Application->doRunCommand(Object(Rector\Core\Console\Command\WorkerCommand
         ), Object(RectorPrefix202207\Symfony\Component\Console\Input\ArgvInput),                                       
         Object(RectorPrefix202207\Symfony\Component\Console\Output\ConsoleOutput))                                     
         #31 vendor/rector/rector/src/Console/ConsoleApplication.php(49):                                               
         RectorPrefix202207\Symfony\Component\Console\Application->doRun(Object(RectorPrefix202207\Symfony\Component\Console\Inp
         ut\ArgvInput), Object(RectorPrefix202207\Symfony\Component\Console\Output\ConsoleOutput))                      
         #32 vendor/rector/rector/vendor/symfony/console/Application.php(208):                                          
         Rector\Core\Console\ConsoleApplication->doRun(Object(RectorPrefix202207\Symfony\Component\Console\Input\ArgvInput),
         Object(RectorPrefix202207\Symfony\Component\Console\Output\ConsoleOutput))                                     
         #33 vendor/rector/rector/bin/rector.php(128): RectorPrefix202207\Symfony\Component\Console\Application->run()  
         #34 vendor/rector/rector/bin/rector(5): require_once('v...')                                                   
         #35 vendor/bin/rector(120): include('v...')                                                                    
         #36 {main}". On line: 284
```

</details>

It looks like it may come from `AddMethodCallBasedStrictParamTypeRector` but if I run Rector only with this rule, error does not happen 🤷🏽‍♂️ 

## Minimal PHP Code Causing Issue

[](https://getrector.org/demo/88bae302-e888-4208-8485-f2f3acaf19dd)

## Expected Behaviour

No error.

### PR DESCRIPTION
**Pull request by samsonasik**
Given the following code:

```php
abstract class SkipIntersectionInUnion
{
    public function prepareParameters(): array
    {
        $parameters =  array_merge($this->getRequestParameters(), [
            'request' => $this->getRequestName(),
        ]);

        $parameters = $this->removeEmptyValues($parameters);

        return array_map(static fn ($parameter) => 'test', $parameters);
    }

    private function removeEmptyValues($input): array
    {
        foreach ($input as &$value) {
            if (! is_array($value)) {
                continue;
            }

            $value = $this->removeEmptyValues($value);
        }

        return array_filter($input, static fn ($val) => true);
    }
}
```

it produce error:

```bash
There was 1 error:

1) Rector\Tests\TypeDeclaration\Rector\ClassMethod\AddMethodCallBasedStrictParamTypeRector\IntersectionTest::test with data set #0 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
Error: Object of class PhpParser\Node\IntersectionType could not be converted to string
```

Fixes https://github.com/rectorphp/rector/issues/7311

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉 /cc @canvural

**Comment by canvural** [comment]
That was super fast. Thank you!

File: rules-tests/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector/FixtureIntersection/skip_intersection_in_union.php.inc [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+            if (! is_array($value)) {
+                continue;
+            }
+
+            $value = $this->removeEmptyValues($value);
+        }
+
+        return array_filter($input, static fn ($val) => true);
+    }
+}
```
Comment (TomasVotruba) [inline rules-tests/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector/FixtureIntersection/skip_intersection_in_union.php.inc]: This looks pretty complex. Is there a way to make this fixture simpler to only contain the intersection type?

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/AddMethodCallBasedStrictParamTypeRector/FixtureIntersection/skip_intersection_in_union.php.inc]: updated with reduce complexity https://github.com/rectorphp/rector-src/pull/2704/commits/dc54a426e029e9020e7a452dc976261718427ea4, that's the only code that reproduce the error

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by samsonasik** [comment]
rebased.

**Comment by samsonasik** [comment]
rebased.

**Comment by TomasVotruba** [comment]
:+1:

**Pull request merged**